modules/apps/transfer/keeper/msg_server.go

```diff
@@ -59,3 +66,29 @@ func (k Keeper) UpdateParams(goCtx context.Context, msg *types.MsgUpdateParams)
 
 	return &types.MsgUpdateParamsResponse{}, nil
 }
+
+// unwindHops unwinds the hops present in the tokens denomination and returns the message modified to reflect
+// the unwound path to take. It ensures only a single token must be present in the tokens list
+// and that the token is not native to the chain.
+func (k Keeper) unwindHops(ctx sdk.Context, msg *types.MsgTransfer) (*types.MsgTransfer, error) {
+	coins := msg.GetCoins()
+	token, err := k.tokenFromCoin(ctx, coins[0])
+	if err != nil {
+		return nil, err
+	}
+
+	if token.Denom.IsNative() {
+		return nil, errorsmod.Wrap(types.ErrInvalidForwarding, "cannot unwind a native token")
+	}
+	var unwindHops []types.Hop
+	for _, trace := range token.Denom.Trace[1:] {
+		unwindHops = append(unwindHops, types.Hop{PortId: trace.PortId, ChannelId: trace.ChannelId}) //nolint: gosimple
+	}
+
+	// Update message fields.
+	msg.SourcePort, msg.SourceChannel = token.Denom.Trace[0].PortId, token.Denom.Trace[0].ChannelId
+	msg.Forwarding.Hops = append(unwindHops, msg.Forwarding.Hops...)
+	msg.Forwarding.Unwind = false
+
+	return msg, msg.ValidateBasic()
```

Comment:
maybe we can return `nil` if `msg.ValidateBasic()` returns an error, otherwise we are returning a modified and invalid message.

Reply:
I think that if an error is returned, the caller should make no assumptions of the other values returned (and shouldn't attempt using it!) but I will make the change anyway :)